shotover/src/server.rs

```diff
@@ -212,11 +291,44 @@ impl<C: CodecBuilder + 'static> TcpCodecListener<C> {
                 crate::connection_span::span(self.connection_count, self.source_name.as_str());
             let transport = self.transport;
 
-            async {
+            let should_return = async {
                 tokio::select! {
-                    // Normal Connection handling
-                    stream_result = Self::accept(&mut self.listener) => {
-                        let stream = stream_result?;
+                    // Wait for a permit to become available and accept new connection
+                    stream_result = async {
+                        // Wait for a permit to become available
+                        let permit = if self.hard_connection_limit {
```

Comment:
why is this logic in the select now?
Can it be moved back to where it was before?